### PR TITLE
Add annotation outlines to Largo thumbnails

### PR DESCRIPTION
**Pull request by lehecht**
- [x] create svg annotations for images
- [x] create svg annotations for videos
- [x] display svgs on top of annotation background
- [x] use css classes to style svgs

---

Added by @mzur:
- [x] A switch to enable/disable the outlines.
- [x] An new flag for the `largo:generate-missing` console command to only generate missing SVGs. The command should also check the SVG existence and regenerate SVGs if they are missing by default.
- [x] Update the manual.
- [x] Run `npm run prod` and commit the files of `src/public/`.
- [x] Also handle whole frame annotations (no SVG needed).
- [x] Add/update tests.

### CODE REVIEW

**Comment by lehecht** [comment]
I forgot to extend the tests. I'm working on it.

**Comment by lehecht** [comment]
I noticed that if you don't set at least the stroke colour, lines and points are not drawn any more. So I need to add it before sending it to the frontend. Or do you have a better idea @mzur ?

**Comment by lehecht** [comment]
> Do you have an example file? So if you don't set the stroke in the SVG, you can't set/add it via CSS later?

Ah, yes. I no longer saw the annotation and thought that the data would also be gone somehow. But thanks for this reminder!

**Comment by lehecht** [comment]
But another question. How should I access the svgs? Currently they are named like the jpg just with the svg ending and are saved in the same directory. 
Quick and dirty would be to use the filesystem/storage. Another idea would be to extend the Image/VideoAnnotation class and database with a xml field for those svgs. Or creating an own class database for the svgs. But maybe there is a simpler solution, @mzur ?

**Comment by mzur** [comment]
> Currently they are named like the jpg just with the svg ending and are saved in the same directory.

That's the way to go, IMO. You can load the SVGs in the same way than the thumbnails (via their public URL, since this storage disk is public).

**Comment by mzur** [comment]
Sometimes easy is also good :smile:

**Comment by lehecht** [comment]
@mzur could you help me with a bug? It has to do with the [rotation angle](https://github.com/biigle/largo/blob/072c35ea2a63f0216cc08e481df8d8f1387ab1cb/src/Jobs/GenerateAnnotationPatch.php#L385-L397) of an ellipse. When you draw an ellipse and use largo, you will see that the ellipse's rotation angle has an small error. Some ellipses are cut like you can see in the picture below. The last picture shows the annotation on the canvas.
Bildschirmfoto vom 2023-12-11 11-38-57
Bildschirmfoto vom 2023-12-11 11-39-14

**Comment by mzur** [comment]
The ellipse is stored with four points like this:

ellipse

I think the thumbnail is just generated based on these points (see red rectangle), so some part of the ellipse could be cut off. I wouldn't put too much effort into getting ellipses right at this point. If there is an issue, we can release the feature as it is and open another GH issue to fix ellipses at some point later. When we looked at the data [two years ago](https://www.frontiersin.org/articles/10.3389/fmars.2021.760036/full) ellipses were only 0.2% of all image annotations.

**Comment by mzur** [comment]
On second thought, you could be right that the rotation is slightly off. Looking at the lower right end of the ellipse, the one in the annotation tool touches the cloud (?) a little differently than the one in the SVG. But whatever the issue is, we can just defer it to some later point.

**Comment by lehecht** [comment]
> On second thought, you could be right that the rotation is slightly off. Looking at the lower right end of the ellipse, the one in the annotation tool touches the cloud (?) a little differently than the one in the SVG. But whatever the issue is, we can just defer it to some later point.

I found out that the container for the thumbnails (flex elements) are smaller than the thumbnails (180px x 135px). Therefore some annotations are cut. Should this be changed?

Bildschirmfoto vom 2023-12-14 08-57-56

File: src/Jobs/GenerateImageAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateImageAnnotationPatch.php:30]: You should use the configured `annotation_format` here.

File: src/Jobs/GenerateVideoAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateVideoAnnotationPatch.php:38]: You should use the configured `annotation_format` here.

**Review by mzur — COMMENTED** [review]
So before I look at the details, some high-level stuff at first:

Why did you choose to send the SVG in the API response and basically rebuild the SVGs with Vue components? Wouldn't it be easier to fetch the SVGs in the same way than the JPEGs and just display them on top of each other?

I see the following slight differences in the shapes of the annotation tool (top) and Largo (middle). I tried to visualize the difference in the bottom row (~yellow is the shape in Largo, ~purple is the annotation).

composite

It's really not much but every shape seems to be moved and/or scaled a bit. This might be related to #121 but then maybe not because the thumbnail and the shape are both displayed inside an SVG.

The only shape that has an actual error is the line string (second from left).

**Comment by lehecht** [comment]
> Wouldn't it be easier to fetch the SVGs in the same way than the JPEGs and just display them on top of each other?

I tried that, but I didn't know how to access/address an object within an image when using the `` to change the style with css.

**Comment by lehecht** [comment]
> The only shape that has an actual error is the line string (second from left).

I see there is something missing, but did you draw the circle or is this also a bug?

**Comment by mzur** [comment]
> I tried that, but later I didn't know how to access/address an object within an image when using the  to change the style with css.

So one the main problem here is lots of data. For example, if there are 10,000 annotations in Largo (which is not unusual), the current implementation would try to load 10,000 files from storage. This would not work. So we need to find some way to implement this that loads only the subset of patches/outlines that are currently visible in the UI.

I get your point that you can't style an SVG that is loaded inside an `` (I didn't think of this before). So maybe we have to fetch the SVG with JavaScript (with `fetch()` or `this.$http.get()`) and then append it manually to the thumbnail element.

> I see there is something missing, but did you draw the circle or is this also a bug?

There is no bug with the circle. I've added that later.

**Comment by mzur** [comment]
I take back my previous comment about the `annotation_format`, by the way. The format should not be configurable. It should be hard-coded (because you cannot change it , really).

Also take my image with the shapes above with a grain of salt. I placed the two images manually on top of each other, so they may not align perfectly. I think it's possible that all the differences are caused by placement of the overlay in the UI. The SVG shape corrdinates may be entirely correct.

**Comment by lehecht** [comment]
> and then append it manually to the thumbnail element.

What exactly do you mean?

**Comment by mzur** [comment]
You can do something like this in the Vue component:

```vue
<template>
<div>
   <img src="mythumbnail.jpg">
   <span v-html="mySvg"></span>
</div>
</template>
<script>
this.mySvg = '<svg>...</svg>';
</script>
```
This will display the JPEG thumbnail and the SVG on top of each other (with the necessary CSS).

**Comment by lehecht** [comment]
So I was able to place the svg on top of the image with your solution and some css, but again I struggle to style it.
I don't know how to access the svg-shapes by using css only. I tried scoped styling (v-deep), but this also didn't work.
Could you help me out again?

**Comment by lehecht** [comment]
Ah I figured out, why nothing worked. Svgs must not contain styling otherwise my approach doesn't work.

**Comment by mzur** [comment]
Usually you can set inline styles or even CSS classes like this:

```html
<style>
.mycircle {
   stroke: red;
   stroke-width: 2;
}
</style>
<svg>
   <circle cx="20" cy="20" r="10" style="fill: blue;" />
   <circle cx="40" cy="40" r="10" class="mycircle" />
</svg>
```

File: src/Jobs/GenerateAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateAnnotationPatch.php:371]: You could also use [SVGPolyline](https://github.com/meyfa/php-svg/blob/main/src/Nodes/Shapes/SVGPolyline.php) (see also [here](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial/Basic_Shapes#polyline)). Then there is no need to return an array any more. Also, the drawn shape will probably look nicer.

File: src/resources/assets/js/components/dismissImageGridImage.vue
Comment (mzur) [inline src/resources/assets/js/components/dismissImageGridImage.vue:77]: Could you also add this directly when the SVG is generated in the backend? Thes would make handling in the frontend easier (no need to parse the XML and convert it back into a string).

File: src/resources/assets/js/components/dismissImageGridImage.vue
Comment (mzur) [inline src/resources/assets/js/components/dismissImageGridImage.vue:74]: You should also handle the case if no SVG is available.

**Review by mzur — CHANGES_REQUESTED** [review]
Looks much better now IMO! I have a few higher-level comments again. I'll have a more detailed look at the code next time.

One thing I noticed is that, while most shapes have a little thicker stroke than in the annotation tool (which is fine), the point is smaller:

image

You can have a look at the annotation tool [styles](https://github.com/biigle/core/blob/master/resources/assets/js/annotations/stores/styles.js) for the point radius.

**Comment by lehecht** [comment]
I changed the point radius to 3, because in the styles.js there is no 'point radius'. The given circle radius looked quiet big to me in comparison to the other shapes, which is why I set the radius to 3. But if you like radius=6 more, then I will change it. I was just not sure.

**Review by mzur — CHANGES_REQUESTED** [review]
The point and line look much better now! I noticed that I can't click on the patches in Largo any more. Probably an event handler has to be moved to a different element (but keep in mind that the SVG may or may not exist here) or the SVG needs `pointer-events: none;` as CSS.

**Comment by mzur** [comment]
I've also added two more todos at the top but we can take care of these when the review of the current code is finished.

**Comment by lehecht** [comment]
> I've also added a few more todos at the top but we can take care of these when the review of the current code is finished.

You added as last point 'handle whole frame annotations'. What do you mean by that? I handled it [here](https://github.com/biigle/largo/blob/6f6169cb41777f245e3134bcf47b443c80a2725a/src/Jobs/GenerateVideoAnnotationPatch.php#L37).

**Comment by mzur** [comment]
> You added as last point 'handle whole frame annotations'. [...] I handled it [here](https://github.com/biigle/largo/blob/6f6169cb41777f245e3134bcf47b443c80a2725a/src/Jobs/GenerateVideoAnnotationPatch.php#L37).

Right, thanks for pointing that out. Then you should update the tests with assertions for the generated SVG files (I added the TODO) and the test for whole frame annotations should check that no SVG is generated.

File: tests/Jobs/GenerateImageAnnotationPatchTest.php
Comment (mzur) [inline tests/Jobs/GenerateImageAnnotationPatchTest.php:32]: This is done like this (otherwise the new ID is not saved in the database):

```suggestion
        $annotation = ImageAnnotationTest::create([
            'shape_id' => Shape::pointId(),
        ]);
```

**Comment by lehecht** [comment]
Let me know, when I can push the new code for the switch and new flags. I'll wait until you approve the last changes.

File: tests/Jobs/GenerateImageAnnotationPatchTest.php
Comment (mzur) [inline tests/Jobs/GenerateImageAnnotationPatchTest.php:301]: Just so you know, there is also `$disk->assertMissing('path')` or `$this->assertNull($content)`. But you don't have to change this here as it works just as well.

File: .gitignore
Comment (mzur) [inline .gitignore:2]: This should not be necessary.
```suggestion
```

File: src/Jobs/GenerateAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateAnnotationPatch.php:6]:
```suggestion
use SVG\SVG;
```

File: src/Jobs/GenerateAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateAnnotationPatch.php:16]:
```suggestion
use SVG\Nodes\Shapes\SVGCircle;
```

File: src/resources/assets/sass/main.scss
Comment (mzur) [inline src/resources/assets/sass/main.scss:102]: Since you already use the scoped style for the dismiss image component, you should put this class there, too (because the name is so generic that it could override some other class.)

File: src/resources/assets/js/components/dismissImageGridImage.vue
Comment (mzur) [inline src/resources/assets/js/components/dismissImageGridImage.vue:8]: This looks like it should rather be:

```
<img @click :src @error>
<div v-if="this.svg !== null" class="overlay">
   <span v-html="this.svg"></span>
   ...
</div>
```

This uses less nesting of HTML elements, does not duplicate the `` and you could even put a `pointer-events: none` in the CSS of the `overlay` so the mouse event on the `` still works with the overlay.

File: src/resources/assets/js/components/dismissImageGridImage.vue
Comment (mzur) [inline src/resources/assets/js/components/dismissImageGridImage.vue:68]: Maybe add an explicit attribute to the point element in the SVG and check this here instead? The current approach is dangerous because if the radius is changed in the backend, it no longer works correctly in the frontend for no obvious reason.

File: src/resources/assets/js/components/dismissImageGridImage.vue
Comment (mzur) [inline src/resources/assets/js/components/dismissImageGridImage.vue:88]: Nice use of CSS variables, I haven't thought of using them this way until now 👍

String concatenation like this can also be easier with JS template strings:

```js
`--color: #${this.image.label_color}; --width: 3px;`
```

**Review by mzur — CHANGES_REQUESTED** [review]
I only have one major comment left (beside the TODOs of above). When the patch is selected, the image fades out to an opacity of `0.25`:

image

This should happen with the SVG, too (i.e. the whole `overlay`).

When this is done, you can proceed with the console command and the settings tab.

**Comment by mzur** [comment]
I just wanted to mention that I will also implement a sorting tab in https://github.com/biigle/largo/pull/120. Since this PR will probably be resolved faster, I will merge the relevant code myself later.

I was hesitant to add additional sidebar tabs in Largo because the tab buttons will be right next to the scrollbar. We'll have to see how well this works UX-wise. We also can't move the sidebar to the left because people are already used to the existing UI.

**Comment by mzur** [comment]
For reference, you can look at https://github.com/biigle/largo/pull/120/commits/16b3328a8aa904e01503587f062789f410831850 to see how to add a sidebar tab.

File: src/resources/assets/sass/main.scss
Comment (lehecht) [reply src/resources/assets/sass/main.scss:102]: If I move this class to the scoped style area, the class does not work any more.
Or did I misunderstand you?

Comment (mzur) [reply src/resources/assets/sass/main.scss]: No that's exactly what I meant. You probably have to add the `::v-deep` there, too.

**Review by mzur — COMMENTED** [review]
All good! Maybe you could add [`stroke-linecap="round"`](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/stroke-linecap) to line string annotations, to make them look more like the ones in the annotation tool.

**Comment by lehecht** [comment]
I have run into a bug, where my button was inactive, but the outline were still visible. (To reproduce: Select label, disable button, deselect label, select random labels).
I found out that the dismissImageGridImage couldn't listen to my event, because it was not mounted yet. So now I emit for every image an event, that the component is mounted and then emit again an event, to change the visibility of the outlines. So there are a lot of events and I'm not happy about that. Otherwise I couldn't find another way to solve this problem. Maybe @mzur you have a very smart idea?

**Review by mzur — COMMENTED** [review]
The issue you describe will be solved if this is implemented in the "Vue way":

Imagine a typical hierarchy of Vue instances (in BIIGLE):

```
- Container
  - Component (image grid)
    - Component (image grid image)
  - Component (sidebar tab)
```

In the "Vue way", information is passed upwards in the hierarchy (e.g. from the sidebar tab to the container) with events. The tab does an `$emit()` and the container listens to the event. The other way around, information is passed downwards in the hierarchy (e.g. from the container to the image grid) as properties of the component. The component defines a prop (e.g. `showOutlines`) and the container sets it dynamically.

And this is exactly how I would implement the information flow whether the outlines should be displayed or not. In cases where you want to pass information from the container "over" an intermediate component (the image grid) into a deeper-level component (the image grid image), you can use [prop drilling](https://vuejs.org/guide/components/provide-inject.html) (the docs are from Vue 3 but it works [the same](https://v2.vuejs.org/v2/api/#provide-inject) with Vue 2 I hope). I haven't used prop drilling in BIIGLE yet, because I learned of this only recently.

The event bus that you currently use should almost never be used in BIIGLE. It's useful when you want to talk from one Vue instance to another on the same page but other than that it's just problematic. If I used it somewhere else and you copied that here, I did it wrong and I'm sorry 😃

Further reading [here](https://medium.com/quick-code/vue-communication-patterns-an-intro-to-props-down-and-events-up-pattern-d53340d2c94).

**Comment by mzur** [comment]
Besides, in my TODO above, I originally meant that you could implement a whole new sidebar tab for "settings". But since this would just increase the conflicts with #120, I like your current placement of the button more! I'll just add the settings tab as part of the merge between #119 and #120.

**Comment by lehecht** [comment]
> you can use [prop drilling](https://vuejs.org/guide/components/provide-inject.html)

This is the solution I was looking for. Thank you!

File: src/resources/assets/js/mixins/largoContainer.vue
Comment (mzur) [inline src/resources/assets/js/mixins/largoContainer.vue:50]: Looks good! I'm curious because I never used prop drilling before. Can't you just return the reactive property of the container instance like this?
```js
return {
   outlines: this.showAnnotationOutlines,
};

Comment (lehecht) [reply src/resources/assets/js/mixins/largoContainer.vue:50]: I tried that, but then outlines is not reactive any more.

Comment (mzur) [reply src/resources/assets/js/mixins/largoContainer.vue:50]: Ok :+1:

**Comment by lehecht** [comment]
I merged the master branch into this one, because I was not sure If I create merge conflicts when merging this branch into the master at the end. This branch was based on an older version of the master and I edited a file there, which was also edited on the newer master branch.

File: src/Console/Commands/GenerateMissing.php
Comment (mzur) [inline src/Console/Commands/GenerateMissing.php:186]: This does not check if the SVG exists first.

File: src/Jobs/GenerateAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateAnnotationPatch.php:372]: Why change this?

File: src/Jobs/GenerateImageAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateImageAnnotationPatch.php]: This would now no longer generate an SVG when an annotation was created, right? But we still need that.

File: src/Jobs/GenerateVideoAnnotationPatch.php
Comment (mzur) [inline src/Jobs/GenerateVideoAnnotationPatch.php]: Same than above, we still need to generate SVGs for new annotations.

**Review by mzur — CHANGES_REQUESTED** [review]
Please read this before you act on my comments below :wink: 

I'm now almost finished with #120 and I would like to wrap up this PR, too. Almost everything that I would ask you to do here, I would have to change or do again during the merge between #120 and this PR. So I'd like to leave most things as they are now.

So here is what I'd like you to do to finish this PR:

- Fix checking for video SVG files in the GenerateMissing command (see comment below).
- Clarify why the point radius was changed (see comment below).
- Revert changes to the GenerateImage/VideoAnnotationPatch jobs and their tests (see comments below).
- Remove the GenerateImage/VideoAnnotationSVGPatch jobs and their tests again. This will be handled differently after the merge (there is a new ProcessAnnotatedImage job which will have flags to ignore patches/feature vectors/SVGs).
- Write a text in this PR that I can paste into the updated manual article.
- Compile and commit the files of `src/public/`.

File: src/Jobs/GenerateAnnotationPatch.php
Comment (lehecht) [reply src/Jobs/GenerateAnnotationPatch.php]: I thought this was a remnant of my point-circle distinction, but I forgot that it was also used to display points with the right size. I reverted it.

**Comment by lehecht** [comment]
Manual: The 'show annotation outlines'-button enables/disables the display of the annotation outlines on the thumbnail.

**Comment by mzur** [comment]
Thanks a lot! This is a huge new feature :+1:

I will take it from here.

**Comment by mzur** [comment]
I decided to backtrack a bit (f7236f4) and display the SVGs as `` (preventing dynamic styling) because:

- The relabel image grid image should show the outlines, too, and an outline color does not make sense there.
- The browser can cache the `` much better and there is no flickering when switching back and forth between the dismiss and relabel steps.
- I found that a dynamically colored outline is not really useful in the dismiss step, too (all patches always have the same color).
- I could position the SVG `` much more precise over the patch `` so the outline position is exact now.

**Review by mzur — APPROVED** [review]
Thanks again for all the work @lehecht!

**Pull request merged**
